tf_helpers/extensions.py

```diff
@@ -0,0 +1,1460 @@
+  entry in `axis`. If `keepdims` is true, the reduced dimensions
+  are retained with length 1.
+  If `axis` has no entries, all dimensions are reduced, and a
+  tensor with a single element is returned.
+  This function is more numerically stable than log(sum(exp(input))). It avoids
+  overflows caused by taking the exp of large inputs and underflows caused by
+  taking the log of small inputs.
+
+  Args:
+    x: The tensor to reduce. Should have numeric type.
+    axis: The dimensions to reduce. If `None` (the default), reduces all
+      dimensions. Must be in the range `[-rank(x), rank(x))`.
+    keepdims: If true, retains reduced dimensions with length 1.
+
+  Returns:
+    The reduced tensor.
+  """
+  return tf_np.asarray(
+      tf.math.reduce_logsumexp(
+          input_tensor=x.data, axis=axis, keepdims=keepdims))
+
+
+def expit(x):
+  """Compute 1 / (1 + exp(-x))."""
+  return tf_np.asarray(tf.math.sigmoid(x.data))
+
+
+def erf(x):
+  """Computes the Gauss error function of x element-wise."""
+  return tf_np.asarray(tf.math.erf(x.data))
+
+
+def _minus(a, b):
+  return [x for x in a if x not in b]
+
+
+def _compose_output_rep(lhs_rep, rhs_rep, lhs_contraction, rhs_contraction,
+                        lhs_batch, rhs_batch):
+  """Compose the output string representation.
+
+  e.g., ij, jk, (((1,), (0,)), ((), ())) -> ik
+        aij, ajk, (((2,), (1,)), ((0,), (0,))) -> aik
+
+  Args:
+    lhs_rep: A string representation for the left-hand side input array
+    rhs_rep: A string representation for the right-hand side input array
+    lhs_contraction: Sequence[int] (the contraction dimensions of lhs)
+    rhs_contraction: Sequence[int] (the contraction dimensions of rhs)
+    lhs_batch: Sequence[int] (the batch dimensions of lhs)
+    rhs_batch: Sequence[int] (the batch dimensions of rhs)
+
+  Returns:
+    A string representation of the result array.
+  """
+  output_rep = []
+  for dim in lhs_batch:
+    output_rep.append(lhs_rep[dim])
+
+  for i in _minus(range(len(lhs_rep)), lhs_batch + lhs_contraction):
+    output_rep.append(lhs_rep[i])
+  for i in _minus(range(len(rhs_rep)), rhs_batch + rhs_contraction):
+    output_rep.append(rhs_rep[i])
+  return "".join(output_rep)
+
+
+def _non_batched_matmul(lhs, rhs, lhs_contraction, rhs_contraction):
+  """Compute the non-batched matrix multiplication.
+
+  If it is the general non-batched/single-batched matrix multiplication,
+  use the highly optimized kernel `tf.tensordot` to handle it.
+
+  Args:
+    lhs: an array (the left-hand side matrix/vector to be multiplied)
+    rhs: an array (the right-hand side matrix/vector to be multiplied)
+    lhs_contraction: Sequence[int] (the contraction dimensions of lhs)
+    rhs_contraction: Sequence[int] (the contraction dimensions of rhs)
+
+  Returns:
+    An array that contains the result.
+  """
+  return tf.tensordot(
+      lhs, rhs, axes=(list(lhs_contraction), list(rhs_contraction)))
+
+
+def tf_dot_general(lhs, rhs, dimension_numbers):
```

Comment:
Should we remove the respective function from `lax.py` since it's now here?

Reply:
Yeah, we can remove it. In that case, we will call `extensions.tf_dot_general` rather than `lax...`, so maybe better keep it?

Reply:
Sure, but in that case I'd recommend just calling tf_dot_general from inside the lax `dot_general` method, to avoid code duplication.

Reply:
Yeah, I will keep everywhere as `lax.dot_general` to make things consistent. Do you mean we should remove `tf_dot_general` inside `extensions`? Thanks!

Reply:
I meant keeping both functions, but removing the function body and helper functions from `lax` and just having there something like `def dot_general(...): return extensions.tf_dot_general(...)`. What do you think?

Reply:
I see, good idea!